Add support for HD6301 and HD6303 microcontrollers

depili opened the pull request:
HD6303 is a Hitachi clone of 6803. This implementation has been done based on the Hitachi HD6301V1/HD6303R User's Manual. The ISA differs from 6805 and 6809.

The flags are implemented as pseudo-registers for better decompilation support. The logic for setting the result flags has been optimized to not produce extra reads from the source address to improve decompilation of volatile data accesses.

I have tested this implementation by reverse-engineering one 8kb firmware image that utilizes most of the opcodes.

depili committed:
Add support for HD6301 and HD6303 microcontrollers

HD6303 is a Hitachi clone of 6803. This implementation has been
done based on the Hitachi HD6301V1/HD6303R User's Manual.

The flags are implemented as pseudo-registers for better decompilation
support. The logic for setting the result flags has been optimized to
not produce extra reads from the source address to improve decompilation
of volatile data accesses.

jobermayr commented:
To fix build errors:
```
diff --git a/Ghidra/Processors/MC6800/certification.manifest b/Ghidra/Processors/MC6800/certification.manifest
index f639a320c4..38ab353b57 100644
--- a/Ghidra/Processors/MC6800/certification.manifest
+++ b/Ghidra/Processors/MC6800/certification.manifest
@@ -21,6 +21,10 @@ data/languages/H6309.slaspec||GHIDRA||||END|
 data/languages/mc146805e2.pspec||GHIDRA||||END|
 data/languages/mc146805e2.slaspec||GHIDRA||||END|
 data/manuals/6809.idx||GHIDRA||||END|
+data/languages/HD6303.cspec||GHIDRA||||END|
+data/languages/HD6303.ldefs||GHIDRA||||END|
+data/languages/HD6303.pspec||GHIDRA||||END|
+data/languages/HD6303.slaspec||GHIDRA||||END|
 data/languages/mc6800.cspec||GHIDRA||||END|
 data/languages/mc6800.ldefs||GHIDRA||||END|
 data/languages/mc6800.pspec||GHIDRA||||END|
```

depili committed:
Add the new files to the manifest

This fixes the build

depili commented:
Thank you, added the files to the manifest in a new commit.

misternebula commented:
Slight error with the opcodes - opcode 0x43 (COM) is listed as being 0x42 in the slaspec, leading to missing decompilation.

depili commented:
Are you talking about the 6809 COM instruction? that is a separate issue and present on the files I didn't touch. This is just adding the HD6303 support.

depili committed:
Bugfix: COM instructions opcode

COM OP3 decoding was erroneus, with the OP_0_3 nibble being decoded as 2 instead of 3. My test firmware didn't use the COM instruction at all so this was undetected.

depili commented:
Oh, sorry, my morning-brain struck. The COM OP3 instruction decoding was indeed erroneus and is now fixed in 22c44b889bbcb3a760c19b994bb1f3c7fc7ad4ba

Thank you for testing.

sigmate commented:
Hi! Using this PR for a HD6303 based reverse-engineering project, it works very nicely, thank you so much. Any update on the merging?
